### PR TITLE
Added PolicyConfigMap and PolicyConfigMapNamespace to KubeSchedulerConfig

### PR DESCRIPTION
**Pull request by whs**

### CODE REVIEW

File: pkg/apis/kops/componentconfig.go [hunk 1]
```diff
@@ -330,6 +330,10 @@ type KubeSchedulerConfig struct {
 	Image string `json:"image,omitempty"`
 	// LeaderElection defines the configuration of leader election client.
 	LeaderElection *LeaderElectionConfiguration `json:"leaderElection,omitempty"`
+	// Name of configmap to use for scheduler policy
+	PolicyConfigMap string `json:"policyConfigMap,omitempty" flag:"policy-configmap"`
```
Comment (chrislovecnm) [inline pkg/apis/kops/componentconfig.go]: Can we re-word the API elements into the typical go syntax for comments?  The first word of the comment is the name, in this case `PolicyConfigMap`.  

> See https://golang.org/doc/effective_go.html#commentary. Comments documenting declarations should be full sentences, even if that seems a little redundant. This approach makes them format well when extracted into godoc documentation. Comments should begin with the name of the thing being described and end in a period:

**Comment by whs** [comment]
- This feature was added in 1.7.0 (kubernetes/kubernetes#43892). Reading that I'll need to add a note that the scheduler must be restarted manually when configmap is modified
- The scheduler will exit if the configmap is not found. I'll add that to the comment, but do we have some way to validate it?
- Could you point to where the docs for kube-scheduler is? I only see kubelet options is documented in the cluster spec and the other cluster services are undocumented. If we don't have one already I can add it.

**Comment by chrislovecnm** [comment]
> This feature was added in 1.7.0 (kubernetes/kubernetes#43892). Reading that I'll need to add a note that the scheduler must be restarted manually when configmap is modified

So we have a bit of a chicken and egg problem with the cluster starting properly. 
@justinsb / @blakebarnett any ideas on how to address this?  If the configmap does not exist, the scheduler will not start.  Without the scheduler starting, cni will not install.  Wonder if we can add a default config map in addons and then have that overriden?  Interesting problem to solve.

@whs the chicken and egg problem is that in a usual case we need to cluster to start.  The cluster depends on the scheduler to get components like CNI running, which is a base component.  Would be optimal if this configmap was reloadable.

We will need to validate that the k8s cluster is over 1.7.x. You can see how we are validating at a component level here: https://github.com/kubernetes/kops/blob/master/pkg/model/components/kubecontrollermanager.go

There is a go file under the same package for the scheduler.

> The scheduler will exit if the configmap is not found. I'll add that to the comment, but do we have some way to validate it?

We can validate that the k8s cluster version is compatible with the cli option.  The chicken and egg problem with k8s being up and the configmap is kinda bugging me.  

> Could you point to where the docs for kube-scheduler is? I only see kubelet options is documented in the cluster spec and the other cluster services are undocumented. If we don't have one already I can add it.

K8s or kops?

**Comment by whs** [comment]
>K8s or kops?

[In this file](https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md#kubelet) only kubelet is documented. Are there other place in kops docs that this option should be documented or should I add to that file?

**Comment by chrislovecnm** [comment]
@whs we do not have full documentation in that markdown file, but when new changes come in we try to add them there.

**Comment by whs** [comment]
If we throw an error when this key is present during cluster creation, only
allowing it with update, would that work?

**Comment by whs** [comment]
All pending concerns should be now resolved, except for that the cluster still won't start with this flag on. Either we do something about that or just document it.

**Comment by justinsb** [comment]
So this PR as it stands is good.  We can map flags, and you've done that perfectly (you've even matched the casing in `pkg/apis/componentconfig/v1alpha1/types.go` in k/k, which is great, because eventually we'll want to reconcile there so if we can avoid differences so much the better).

So the question is where should the configuration live, I guess.  Is there a default scheduler configuration we could create?  (just as we create a default LimitRange, for example: https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/resources/addons/limit-range.addons.k8s.io/v1.5.0.yaml ).  This feels like the first step towards componentconfig, which is super exciting.

I'm guessing what we should do is to automatically create a default scheduler configmap (perhaps in 1.9), and point the scheduler to read from that configmap.  We'd have to make sure we didn't replace it on kops/kubernetes updates I guess.   Does that sound right @whs ?

/lgtm

**Comment by whs** [comment]
The default config for kube-scheduler is hardcoded in [scheduler's source](https://github.com/kubernetes/kubernetes/blob/31697a27f9e7ff5f7f80f61955106d2f33e9165b/plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go#L123). It should be like [this](https://gist.github.com/whs/3381fb2e3e17c672ebe99936b1f661b6) if written as JSON.

I'm not sure that providing a default scheduler configmap would be the best solution here as we would have to keep track of scheduler default policy update, and update the file. (for example, when pod affinity it was implemented by adding a predicate and priority) Also, it leads to the question on what should we do if the user updated the file and upgrade the cluster.

To be clear, if the user doesn't provide `--policy-configmap` the scheduler would just start with the default configuration. The only problem is when this option is present, but the referenced configmap is not present.

Providing default configmap if the user specified this option is one way we could make this work, but it would be complicated. Since this is an option for advanced users, I think noting in the docs and godoc that this option will not work on cluster creation should be fine.

Is this ok with you? And is it possible to detect cluster creation when validating the spec so I can add this check in the code as well?

**Comment by whs** [comment]
I added the docs. I think we're ready to merge.

**Comment by justinsb** [comment]
/lgtm

Hopefully they'll add reloading once it becomes clearer how reloading should work :-)

**Comment by chrislovecnm** [comment]
@whs something I thought about.  We could have kops manage the config map.

**Comment by whs** [comment]
That could work, and was pointed out by @justinsb earlier as well. As this option is optional (if you don't specify anything the scheduler have its hardcoded default), I think adding kops-managed scheduler policy would add more burden to the kops maintainers, as we need to keep track of upstream policy changes.

**Comment by chrislovecnm** [comment]
@whs only create the config map if the user wants it.  And set the scheduler flag.  We have dynamic addons.  Let me know if you are interested in how to do this.

**Comment by whs** [comment]
I think I have some idea now. Give me an hour..

**Pull request merged**